### PR TITLE
Grouped assessment query

### PR DESCRIPTION
**Pull request by gigxz**
https://www.pivotaltracker.com/story/show/184882775

### CODE REVIEW

File: drivers/hmis/app/models/hmis/hud/custom_assessment.rb [hunk 1]
```diff
@@ -155,6 +165,33 @@ def self.new_with_defaults(enrollment:, user:, form_definition:, assessment_date
...
+      # If we have a source, find annuals "near" it (within threshold)
+      # If we don't have a source, that means this is a new annual. Include any annuals from the past 3 months.
+      source_date = source_assessment&.assessment_date || Date.current
+      household_assessments.group_by(&:personal_id).
+        map do |_, assmts|
+          nearest_assmt = assmts.min_by { |a| (source_date - a.assessment_date).abs }
+          distance = (source_date - nearest_assmt.assessment_date).abs
+
+          nearest_assmt if distance <= threshold
+        end.compact
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/hud/custom_assessment.rb:189]: Thoughts on considering assessments to be the "same" annual if they were performed within 3 months of the annual that the user opened? We are not grouping annuals together in the database.

I haven't yet worked out how annuals will work in the UI in terms of setting the information date. Right now you have to input the information date individually for each client.

File: drivers/hmis/app/models/hmis/hud/custom_assessment.rb [hunk 2]
```diff
@@ -155,6 +165,33 @@ def self.new_with_defaults(enrollment:, user:, form_definition:, assessment_date
...
+    source_assessment = Hmis::Hud::CustomAssessment.find(assessment_id) if assessment_id.present?
+    # FIXME wont work for wip.
+    # raise HmisErrors::ApiError, 'Assessment not in household' if source_assessment.present? && !enrollments.pluck(:enrollment_id).include(source_assessment.enrollment_id)
+
+    household_assessments = Hmis::Hud::CustomAssessment.with_role(assessment_role.to_sym).for_enrollments(household_enrollments)
+
+    case assessment_role.to_sym
+    when :INTAKE, :EXIT
+      # Ensure we only return 1 assessment per person
+      household_assessments.group_by(&:personal_id).map { |_, a| a.last }
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/hud/custom_assessment.rb]: any reason not to just:
```suggestion
      household_assessments.index_by(&:personal_id).values
```

File: drivers/hmis/app/models/hmis/hud/custom_assessment.rb (hunk 1 above)
Comment (eanders) [reply drivers/hmis/app/models/hmis/hud/custom_assessment.rb:189]: I think that's a reasonable window.  As we discussed, this doesn't actually change any dates, just groups things together in the display for editing.

**Review by eanders — APPROVED** [review]

File: drivers/hmis/app/models/hmis/hud/custom_assessment.rb (hunk 2 above)
Comment (gigxz) [reply drivers/hmis/app/models/hmis/hud/custom_assessment.rb]: nope, thats better, thanks!

**Pull request merged**